### PR TITLE
Provide the CMake install the necessarily information (and flags) to hand Windows DLL and add Windows installation tests

### PR DESCRIPTION
**Pull request by lemire**
Currently, under Windows, users can build and install a library (static or dynamic). If they used a dynamic library, they need to manually set SIMDJSON_USING_LIBRARY to 1, or else they will experience linkage errors.

This cannot be avoided in general because Windows (or is it just Visual Studio) requires programmers to change their code depending on whether the linkage is dynamic or static.

However, if the user relies on CMake, then we already have the opportunity to provide them with some flags, if they import the dependency as a CMake dependency.

This is, in my view, a less invasive solution compared to the solution proposed in https://github.com/simdjson/simdjson/pull/1455 by @JackBoosY because we do rely on CMake so much. This is important because not all of our users rely on CMake.

Related to https://github.com/microsoft/vcpkg/pull/16105

### CODE REVIEW

**Comment by lemire** [comment]
@JackBoosY Can you review my alternative approach?

File: include/simdjson/common_defs.h [hunk 1]
```diff
@@ -149,14 +149,27 @@ constexpr size_t DEFAULT_MAX_DEPTH = 1024;
...
+    #if SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY
+    // Windows user who call a dynamic library should set SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY to 1.
     #define SIMDJSON_DLLIMPORTEXPORT __declspec(dllimport)
-    #else
+    #elif SIMDJSON_BUILDING_WINDOWS_DYNAMIC_LIBRARY
+    // We set SIMDJSON_BUILDING_WINDOWS_DYNAMIC_LIBRARY when we build a DLL under Windows.
     #define SIMDJSON_DLLIMPORTEXPORT __declspec(dllexport)
+    #else
+    // We assume by default static linkage
+    #define SIMDJSON_DLLIMPORTEXPORT
```
Comment (JackBoosY) [inline include/simdjson/common_defs.h:172]: For dynamic build, we can only use `dllexport` by setting macro `SIMDJSON_BUILDING_WINDOWS_DYNAMIC_LIBRARY` at build time, without exporting the macro `SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY` during use.
For static builds, we always need to define static macro, so we still need to change the header file downstream.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by lemire** [comment]
@JackBoosY Can you elaborate? 

Let me explain what I think should happen with this PR...

1. If you are building the DLL, then you set SIMDJSON_BUILDING_WINDOWS_DYNAMIC_LIBRARY to 1 (but not SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY) and it triggers  __declspec(dllexport). You are good.
2. If you are using the DLL, then you set SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY to 1 and it will trigger __declspec(dllimport) which are needed. You are good.
3. If you had a static build, then neither of these variables are set and you get neither  __declspec(dllexport) nor __declspec(dllimport), and you are good.

After the library is installed then either SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY is set to 1, or not. This gets automatically exported in the simdjson-config.cmake file.

Let me consider your comment...

> For dynamic build, we can only use dllexport by setting macro SIMDJSON_BUILDING_WINDOWS_DYNAMIC_LIBRARY at build time, without exporting the macro SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY during use.

This PR is based on the assumption that you need the dllimport and so you need to set SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY if you are going to link against the DLL... but that's fine because we export SIMDJSON_USING_WINDOWS_DYNAMIC_LIBRARY=1 in  simdjson-config.cmake when it is needed (when a DLL was built).

Am I wrong?

> For static builds, we always need to define static macro, so we still need to change the header file downstream.

I do not understand why you say that you need to change the header file.

**Comment by lemire** [comment]
With the latest commits, we fully test the installation, under Windows, in both static and dynamic mode.

**Comment by lemire** [comment]
@JackBoosY @jkeiser Took me some time, but I can now verify that you can automatically link with both a DLL and a static lib. under Windows from the CMake install using find(simdjson). There is no need for the user of the library to set anything it should all be automated. Importantly, it requires no modification to the source code during installation.

**Comment by JackBoosY** [comment]
Our final goal is to define as few macros as possible when using simdjson, but the current implementation is not like this.
 - For using it through cmake, the modification is fine now.
 - For the integration of vcpkg into the msvc project, we still need to change the header file according to build type(dynamic/static). This is because vcpkg's integration of msvc can NOT automatically add these macros.

**Comment by JackBoosY** [comment]
In my modification, these files that define `SIMDJSON_DLLIMPORTEXPORT` are generated according to the build type.
So:
 - When it is statically built, `SIMDJSON_DLLIMPORTEXPORT` is always empty regardless of whether it is built or used. 
 - When it is built dynamically, we need to add macro during the build process to export these functions, and we can import these function symbols without adding any macros during use.

**Comment by lemire** [comment]
If vcpkg is going to keep on patching the source, then should I keep the following lines around:

```C++
// workaround for vcpkg
#if SIMDJSON_USING_LIBRARY
#define SIMDJSON_DLLIMPORTEXPORT __declspec(dllimport)
#endif 
```

I am otherwise replacing `SIMDJSON_USING_LIBRARY` by a macro that is clearer since I feel that `SIMDJSON_USING_LIBRARY` is not very clear... But it would be fine to keep it in a corner just for vcpkg.

>For the integration of vcpkg into the msvc project, we still need to change the header file according to build type(dynamic/static). This is because vcpkg's integration of msvc can NOT automatically add these macros.

Ok. So vcpkg  will also fail to pick up `SIMDJSON_THREADS_ENABLED=1` ? Or do you also have a patch for that ?

If that's how vcpkg works, that it, it does not pick up CMake MACRO configuration from the installation, then, yeah, you have no choice but to modify the source code. I am sorry for taking so long to understand... it is not intuitive to me that vcpkg would work that way, but if it is, then the issue is indeed not too complex.


Thanks for the feedback and for the pull request. Your pull request helped me greatly. :-)

**Comment by lemire** [comment]
@JackBoosY I added the `#if SIMDJSON_USING_LIBRARY` workaround so your patch should keep on working.

**Pull request merged**